allow empty --prompt-cache file

divinity76 opened the pull request:
This allows the use of mkstemp(), std::tmpfile(), Python's tempfile.NamedTemporaryFile(), and similar create-empty-file API's for the user.

I switched from the C fopen API to the C++ filesystem api to get around the fact that, to the best of my knowledge, C has no portable way to get the file size above LONG_MAX, with std::ftell() returning long? fallback to std::ifstream for c++  < 17 (the project is currently targeting C++11 it seems - file_exists() and file_size() can be removed when we upgrade to c++ >= 17)

divinity76 committed:
allow empty --prompt-cache file

This allows the use of std::tmpnam(), std::tmpfile(), Python's tempfile.NamedTemporaryFile(), and similar create-empty-file API's for the user.

I switched from the C fopen API to the C++ filesystem api to get around the fact that, to the best of my knowledge, C has no portable way to get the file size above LONG_MAX, with std::ftell() returning long? fallback to std::ifstream for c++  < 17
(the project is currently targeting C++11 it seems - file_exists() and file_size() can be removed when we upgrade to c++17)

ggerganov commented on examples/main/main.cpp:
```diff
@@ -215,24 +237,27 @@ int main(int argc, char ** argv) {
             session_tokens.resize(n_ctx);
             size_t n_token_count_out = 0;
-            if (!llama_load_session_file(ctx, path_session.c_str(), session_tokens.data(), session_tokens.capacity(), &n_token_count_out)) {
+            if (!llama_load_session_file(ctx, path_session.c_str(), session_tokens.data(), session_tokens.capacity(), &n_token_count_out))
+            {
```
Opening braces on same line:

```suggestion
            if (!llama_load_session_file(ctx, path_session.c_str(), session_tokens.data(), session_tokens.capacity(), &n_token_count_out)) {
```

divinity76 replied:
thanks, formatted: 0fcdefc346d584d05c4cc7486bc87cc02b097763

divinity76 committed:
formatting

(requested in codereview)

slaren commented on examples/main/main.cpp:
```diff
@@ -10,6 +10,9 @@
 #include <cstring>
 #include <ctime>
 #include <fstream>
+#if __cplusplus >= 201703L
+#include <filesystem>
+#endif
```
Considering that we build for C++11, is this ever going to be true?

divinity76 replied:
not anytime soon. I hope we will not target C++11 forever though? c++17 is already 7 years old

slaren replied:
Maybe so, but that's a different discussion. I don't see the point of including code that is never going to be used.

divinity76 commented:
> Hm, why do you need uint64_t file sizes? You seem to be checking if the file is empty or not?

I played with a session file 1.8GB in size, that is dangerously close to MSVC's LONG_MAX (2GB), a 64bit data type is warranted. probably irrelevant if its uint64_t or int64_t or size_t or ssize_t though. feel free to change it

I don't actually know what MSVC's std::ftell() returns when it's over LONG_MAX , does it return 0? LONG_MAX? something else? i have no idea!

ggerganov reviewed (changes requested):
So I got a bit confused earlier. Let's remove all the C++17 stuff and implement a regular `bool file_is_empty()` function since this is all we need

cebtenzzre commented:
> C has no portable way to get the file size above LONG_MAX, with std::ftell() returning long?

Unix has ftello, and off_t will be 64 bits everywhere except glibc where you have to use `-D_FILE_OFFSET_BITS=64`. Windows has _ftelli64.

divinity76 committed:
remove c++17, file_is_empty

divinity76 commented:
@slaren @ggerganov  ok, i removed the c++17 stuff and replaced file_size() with file_is_empty()

@cebtenzzre yeah but we have the luxury of C++ APIs supporting 64bit, so there is no need to resort to OS-specific APIs :)
